e2e/stack_monitoring_test.go

```diff
@@ -0,0 +1,614 @@
+}
+
+// StackMonitoringFeatureContext adds steps to the Godog test suite
+func StackMonitoringFeatureContext(s *godog.Suite) {
+	testSuite := StackMonitoringTestSuite{
+		Env:            map[string]string{},
+		collectionHits: map[string]map[string]interface{}{},
+	}
+
+	s.Step(`^"([^"]*)" sends metrics to Elasticsearch using the "([^"]*)" collection monitoring method$`, testSuite.sendsMetricsToElasticsearch)
+	s.Step(`^the structure of the documents for the "([^"]*)" and "([^"]*)" collection are identical$`, testSuite.checkDocumentsStructure)
+
+	s.BeforeSuite(func() {
+		log.Debug("Before StackMonitoring Suite...")
+
+		env := map[string]string{
+			"stackVersion":              stackVersion,
+			"xpackMonitoringCollection": "false",
+		}
+
+		log.Debug("Installing elasticsearch monitoring instance")
+		startRuntimeDependencies("stack-monitoring", env, 3)
+	})
+	s.BeforeScenario(func(*messages.Pickle) {
+		log.Debug("Before StackMonitoring Scenario...")
+	})
+	s.AfterSuite(func() {
+		log.Debug("After StackMonitoring Suite...")
+		log.Debug("Destroying elasticsearch monitoring instance, including attached services")
+		tearDownRuntimeDependencies("stack-monitoring")
+	})
+	s.AfterScenario(func(*messages.Pickle, error) {
+		log.Debug("After StackMonitoring Scenario...")
+		testSuite.removeProduct()
+	})
+}
+
+// assertHitsEqualStructure returns an error if hits don't share structure
+func assertHitsEqualStructure(sm *StackMonitoringTestSuite, collectionMethod1 string, collectionMethod2 string) error {
+	collectionMethod1Hits := sm.collectionHits[collectionMethod1]
+	collectionMethod2Hits := sm.collectionHits[collectionMethod2]
+
+	collectionMethod1HitsJSON := gabs.Wrap(collectionMethod1Hits)
+	collectionMethod2HitsJSON := gabs.Wrap(collectionMethod2Hits)
+
+	err := checkParity(sm, collectionMethod1HitsJSON, collectionMethod2HitsJSON)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func checkParity(sm *StackMonitoringTestSuite, legacyContainer *gabs.Container, metricbeatContainer *gabs.Container) error {
+	allowedInsertionsInMetricbeatDocs := []string{
+		"service",
+		"@timestamp",
+		"agent",
+		"event",
+		"host",
+		"ecs",
+		"metricset",
+	}
+	allowedInsertionsInMetricbeatDocs = append(allowedInsertionsInMetricbeatDocs, sm.allowedInsertionsExtra...)
+
+	allowedDeletionsFromMetricbeatDocs := []string{
+		"source_node",
+	}
+	allowedDeletionsFromMetricbeatDocs = append(allowedDeletionsFromMetricbeatDocs, sm.allowedDeletionsExtra...)
+
+	hitsPath := "hits.hits"
+	legacyHits := legacyContainer.Path(hitsPath)
+	metricbeatHits := metricbeatContainer.Path(hitsPath)
+
+	legacyTypes, legacySources := checkSourceTypes(legacyHits)
+	metricbeatTypes, metricbeatSources := checkSourceTypes(metricbeatHits)
+
+	if len(legacyTypes) > len(metricbeatTypes) {
+		// returns an array as the result of removing all elements in 'b' from the 'a' array
+		// only used here
+		arrayDiff := func(a []string, b []string) []string {
+			target := map[string]bool{}
```

Comment:
Fun fact: `bool` takes more memory (1 byte) than an empty struct (0 bytes)! See https://dave.cheney.net/2014/03/25/the-empty-struct.

So when using a map for this sort of "existence-check" pattern, where we don't really care about the value of the map entries, consider using `map[string]struct{}`.